### PR TITLE
CAS-1057 Bug: Edit message when offline

### PR DESCRIPTION
**Pull request by samiuelson**
https://stream-io.atlassian.net/secure/RapidBoard.jspa?rapidView=81&projectKey=CAS&selectedIssue=CAS-1103

### 🎯 Goal
Allow editing message offline without issues.

There are multiple async operations executed in the `ChatDomainImp::connectionRecovered` method. Those operations affect a single DB table `stream_chat_message`. Resulting in unpredictable and wrong behavior related to race conditions. 

In ~80% of cases message offline message editing fails because before `::retryEntities` method is called from the `::connectionRecovered` method the edited message is overridden in DB already by the async logic intended to update Channel data.

This PR fixes those race conditions and synchronizes `ChatDomain::retryFailedEntities` suspending function.

### 🛠 Implementation details

1. Remove race conditions caused by async operations performed by `ChatDomainImpl::connectionRecovered` method, where the shared resource is `DB.stream_chat_message` table.
2. Synchronize `ChatDomainImpl::retryFailedEntities` method using `Mutex`. It's called from `::connectionRecovered` method and from `HealthEvent` and we're not guaranteed it's called on the same thread.

### 🎨 UI Changes

| Before | After |
| --- | --- |
|<video src="https://user-images.githubusercontent.com/4527432/124447813-0851d600-dd82-11eb-818b-427213872d8b.mp4"/>|<video src="https://user-images.githubusercontent.com/4527432/124447847-10aa1100-dd82-11eb-9f75-dd795e8c1418.mp4"/>|

### 🧪 Testing

When the message is edited while the device is offline, after switching to the online message should not be reverted to the old one.

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (CMS, cookbooks, tutorial)
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt [hunk 1]
```diff
@@ -681,23 +690,21 @@ internal class ChatDomainImpl internal constructor(
...
-        // 4 retry any failed requests
-        if (online) {
-            retryFailedEntities()
-        }
     }
 
+    // An attempt to synchronize ::retryFailedEntities execution since it is called from multiple places. The shared resource is DB.stream_chat_message table.
+    private val entitiesRetryMutex = Mutex()
     internal suspend fun retryFailedEntities() {
-        delay(1000)
```
Comment (samiuelson) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:695]: Good bye, old friend!

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt [hunk 2]
```diff
@@ -101,6 +101,9 @@ internal class EventHandlerImpl(
                     }
                 }
                 is HealthEvent -> {
+                    // TODO: consider removing caling the code below on every HealthCheckEvent
+                    // We have retry logic for the most crucial network calls. Why would we need to rely on another mechanism for doing retries?
+                    // We already retry entities inside `domainImpl.connectionRecovered` which is called when the ConnectedEvent is received ☝️
                     domainImpl.scope.launch {
```
Comment (samiuelson) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt:107]: Can we remove this call and do nothing on HealthEvent as described ? 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt [hunk 3]
```diff
@@ -681,23 +689,21 @@ internal class ChatDomainImpl internal constructor(
...
         }
-
-        // 4 retry any failed requests
-        if (online) {
-            retryFailedEntities()
-        }
     }
 
+    // Synchronizing ::retryFailedEntities execution since it is called from multiple places. The shared resource is DB.stream_chat_message table.
+    private val entitiesRetryMutex = Mutex()
```
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:698]: Maybe move to fields declaration place?

**Review by ogkuzmin — APPROVED** [review]
Good job, Sam!

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**